### PR TITLE
API maturity annotation

### PR DESCRIPTION
**Pull request by tombentley**

### CODE REVIEW

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
```
Comment (tombentley) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java:5]: I stole these annotations from Apache Kafka, but have changed them slightly (with the aim of better suiting our needs). I don't think it's entirely right to assert copyright over them, even though I understand that such copyright assertions aren't necessarily meaningful from a (US) legal PoV.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/package-info.java [hunk 2]
```diff
@@ -38,4 +38,5 @@
  * the communication between the client and broker.</p>
  *
  */
+@ApiMaturity.Unstable
```
Comment (tombentley) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/package-info.java:41]: Just mark everything as unstable for now.

File: pom.xml [hunk 3]
```diff
@@ -277,6 +277,9 @@
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-javadoc-plugin</artifactId>
                     <version>3.6.0</version>
+                    <configuration>
+                        <breakiterator>true</breakiterator>
```
Comment (tombentley) [inline pom.xml:281]: This makes the Javadoc "first sentence is a summary" convention work in the presence of `i.e.` etc.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java (hunk 1 above)
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java:5]: Given the build fails without us asserting copyright it seems necessary to hold your nose and do so.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    @Target({ ElementType.TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.PACKAGE })
+    @Retention(RetentionPolicy.SOURCE)
+    @Documented
+    public @interface Evolving {
+    }
+
+    /**
+     * Compatibility is maintained in major, minor and patch releases with one exception: compatibility may be broken
+     * in a major release (i.e. 0.m) for APIs that have been deprecated for at least one major/minor release cycle.
```
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java:46]: I'm not quite sure how to read `at least one major/minor release cycle.` 

Does it mean that a the old method gets deprecated in 0.4 are removed 0.5? Or 0.6 (i.e. they survive for one full minor cycle)? Or do you mean deprecated in 0.4 and can only be removed in 1.0 ? 

I think an examples table would be in order here.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/package-info.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/package-info.java:41]: HAHAHA, I really thought you were pushing for at least evolving for filter API.

**Review by SamBarker — COMMENTED** [review]
I'm happy with the three levels outlined but I'm struggling to align the contract outlined in the annotations with [SEMVER 2.0](https://semver.org/). 

I'd also add [Major Versions are not sacred](https://tom.preston-werner.com/2022/05/23/major-version-numbers-are-not-sacred) to the reading pile. As I admit I was guilty of thinking of them as being much more significant than they really are.

I'm +1 assuming the contract is semver compliant.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/package-info.java (hunk 2 above)
Comment (tombentley) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/package-info.java:41]: Just getting the mechanism for evolution in place first. Actually changing the annotations would be a whole other  discussion.

**Comment by tombentley** [comment]
You raise good points. In general I'm fine with SemVer and I'm fine with not attaching any particular meaning to major versions. I think it's question of _when_ we start applying those rules, what we apply those rules _to_, and the _consequences_ of those rules. 

With semver it's a whole released artifact that gets a version number. For us this means a whole module (e.g. the api module). Before/after 1.0 then either nothing has a compat guarantee or everything does. Within the rules set out by semver all we can do to avoid this is split up the api module. But we don't want to do that because it's burdensome. 

The idea with the annotations (though maybe not _these_ annotations) is they would allow us to operate at a finer-grained level than a whole released artifact. We could opt to stabilise some bits of API within a module (e.g. filters, filter context) in 0.x versions, and do so before others (e.g. filter factories). The beneficiary of this, during 0.x, wouldn't be the filter author (who has to use whatever API is available if they're going to implement a given functionality), but ourselves (recording what we think is done, and what is not done). Clearly everything would need to be stable before we declared 1.0, and at that point we could either choose to remove the annotations entirely, or keep them for declarations we specifically want to remove from the guarantees provided by semver.

Strictly speaking if we used semver and defined "compatible" as binary and source compatible, and never messed up then we could increment api module versions in lockstep with the runtime. But practically speaking I think it's better for the filter API versions to move more slowly.  We have other public interfaces than the filter API (for example the rest of config file outside of the `filters` is also a public interface). We might want to iterate those more quickly than the filter API. This point obviously doesn't apply to this PR, but I think it's worth discussing in this context.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java (hunk 4 above)
Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/ApiMaturity.java:46]: This is my reading.

## release timeline 1:
1. 0.4 deprecates methodA
2. 0.5 release made
3. 1.0 release made - removes methodA

## release timeline 2
1. 0.4 deprecates methodA
2. 1.0 release made - cannot remove methodA
3. 1.1 release made - still cannot remove methodA
4. 2.0 release made - methodA removed

**Comment by SamBarker** [comment]
> With semver it's a whole released artifact that gets a version number. For us this means a whole module (e.g. the api module). Before/after 1.0 then either nothing has a compat guarantee or everything does. Within the rules set out by semver all we can do to avoid this is split up the api module. But we don't want to do that because it's burdensome.

I think we approach the implications of semver differently. I think we agree on the mechanics however. So I interpret semver as saying changes which:
- do not alter any of the public API increment patch
- **are** backwards compatible in the public API increment minor
- are **not** Backwards compatible in the public API increment major

but the implication of that is Major needs to loose any other meaning. I.e. its a nudge to users that moving from 1.X to 2.X requires work but moving from 1.X to 1.Y is safe. (I'm deliberately ignoring the 0.X releases here)

Within that public API surface we can then use annotations to make differing commitments. I.e we could annotate the generated Request and response interfaces as `Stable` and thus users know that we do not expect these to change. We could in the same semver version have `Filter` marked as `Evolving` which I think means we think its stable and any changes will be backwards compatible. Filter authors then get a sense of reassurance that are programming against a stable API surface. `@Unstable` we make no commitments. 

However we would still increment the major version when we released changes breaking changes to an `@Unstable` API.

So if we adopt semver the stability annotations might not have a technological impact but they serve as reminders for the dev team of what refactoring options they have and as signals to users of our APIs.

**Comment by tombentley** [comment]
@SamBarker that's a good summary of what I'm suggesting. 

I'm not really sure that there's a worthwhile distinction between `@Stable` and `@Evolving` under this scheme, is there? Perhaps we only need two annotations: `@SemVer` (semantics as per semver for whatever release artifact this ends up in) and `@Unstable` (all bets are off)?

**Comment by SamBarker** [comment]
>I'm not really sure that there's a worthwhile distinction between @Stable and @Evolving under this scheme, is there?
/shrug

Can I play the `It Depends` ™️ card? I suspect there is no tangible benefit, but there is a psychological one. They represent different levels of confidence and thus reliability.

I think I'm also arguing that nothing in 0.x could be marked `@stable` but could be marked `@evolving`. Maybe moving to 1.0 would happen when we are ready to promote `@evolving` to `@stable` ?

**Comment by SamBarker** [comment]
I've opened https://github.com/kroxylicious/kroxylicious/pull/645 to explore the idea of enforcing semver compliance in our API module. At this stage I'm not taking any positions as to how that relates to this discussion.

**Comment by robobario** [comment]
Reviewed Jan 31 2024, call to have another look and consider enforcing semver vs these annotations

**Pull request closed without merging**